### PR TITLE
*: Upgrade rust toolchain to nightly-2023-12-10 

### LINKED ISSUE #15581: update rust toolchain
## Development Task

The currently rust-toolchain `nightly-2022-11-15` is too old we should update it to a recent version.

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Close #15581

What's Changed:

```commit-message
Upgrade rust toolchain to nightly-2023-12-10
```

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
 See https://github.com/tikv/tikv/pull/16213#issuecomment-1874915199.

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by overvenus** [comment]
The performance before and after this change is almost the same.

Benchmark results:

| Commit | TPCC | sysbench |
|--------|--------|--------|
| 4702b9bf6ed8d72d20bedf3de184f63111055b8e | image | image |
| e3702be579323bb145729d742618f49375656129 | image | image |

File: components/resource_control/src/worker.rs [hunk 1]
```diff
@@ -332,10 +332,13 @@ impl<R: ResourceStatsProvider> PriorityLimiterAdjustWorker<R> {
...
-            .get_priority_resource_limiters()
-            .zip(TaskPriority::priorities())
-            .map(|(l, p)| PriorityLimiterStatsTracker::new(l, p.as_str()));
+        let limiters = resource_ctl.get_priority_resource_limiters();
+        let priorities = TaskPriority::priorities();
+        let trackers = [
+            PriorityLimiterStatsTracker::new(limiters[0].clone(), priorities[0].as_str()),
+            PriorityLimiterStatsTracker::new(limiters[1].clone(), priorities[1].as_str()),
+            PriorityLimiterStatsTracker::new(limiters[2].clone(), priorities[2].as_str()),
+        ];
```
Comment (glorv) [inline components/resource_control/src/worker.rs:341]:
```suggestion
        let trackers = std::array::from_fn(|i| PriorityLimiterStatsTracker::new(limiters[i].clone(), priorities[i].as_str()));
```

**Review by LykxSassinator — APPROVED** [review]

File: tests/failpoints/cases/test_engine.rs [hunk 2]
```diff
@@ -63,20 +63,15 @@ fn test_write_buffer_manager() {
...
-// 5.                                           pick memtables to flush:
-//                                            t1(0-10), t2(10-20)
-//                                            flush job(0-20)
-// 6.                                           finish flush
-// 7.                                           unlock
-// 8.                                           on_flush_completed:
-//                                            update last_flushed to 20
-// 9.       on_memtable_sealed
-//        10 > 20 *panic*
+// 1. lock
```
Comment (glorv) [inline tests/failpoints/cases/test_engine.rs:66]: Better skip format this comment

File: components/raftstore-v2/src/operation/command/admin/transfer_leader.rs [hunk 3]
```diff
@@ -50,21 +50,21 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
-    ///     Leader will send a MsgTransferLeader to follower.
-    /// 2. execute_transfer_leader on follower
-    ///     If follower passes all necessary checks, it will reply an
-    ///     ACK with type MsgTransferLeader and its promised applied index.
-    /// 3. ready_to_transfer_leader on leader:
-    ///     Leader checks if it's appropriate to transfer leadership. If it
-    ///     does, it calls raft transfer_leader API to do the remaining work.
+    /// 1. pre_transfer_leader on leader: Leader will send a MsgTransferLeader
+    ///    to follower.
+    /// 2. execute_transfer_leader on follower If follower passes all necessary
```
Comment (glorv) [inline components/raftstore-v2/src/operation/command/admin/transfer_leader.rs:55]:
```suggestion
    /// 2. execute_transfer_leader on follower: If follower passes all necessary
```

**Review by glorv — APPROVED** [review]
LGTM, a few nit coments

**Pull request merged**